test/object-store/sync/flx_sync.cpp

```diff
@@ -4927,6 +4927,255 @@ TEST_CASE("flx: nested collections in mixed", "[sync][flx][baas]") {
     CHECK(nested_list.get_any(1) == "foo");
 }
 
+TEST_CASE("flx: role change bootstrap", "[sync][flx][baas][role_change][bootstrap]") {
```

Comment:
just my 2 cents, but I think the test(s) can be simplified a great deal by doing the following:

- instead of using `register_connection_change_callback` you can use the debug hook and check the client receives a 200 error
- for query version 1 you can count all non-empty download messages (the role change bootstrap essentially) and assert on that (additionally, once all messages are received (last_in_batch=true for BootstrapMessageProcessed) you can check the PendingBootstrapStore stored the right data)
- finally, you assert the expected data is in the realm
- if you want multiple bootstrap in a single test, you can optionally pause the session, change the roles, and resume the session to test role changes detection at ident

Reply:
didn't mention the state machine, but I'm not sure you need it, do you?

Reply:
The state machine is being used for two reasons:
* Wait for the 200 error to be received
* Verify the order of receiving a server initiated bootstrap with multiple messages, so we track the order/steps of MoreToCome, LastInBatch and then bootstrap processed.

Reply:
Isn't wait_for_download enough for all of this? and you can do the tracking in `on_sync_client_event_hook`